src/main/java/com/iota/iri/service/garbagecollector/GarbageCollector.java

```diff
@@ -0,0 +1,323 @@
+     */
+    final SnapshotManager snapshotManager;
+
+    /**
+     * Holds a reference to the {@link TipsViewModel} which is necessary for removing tips that were pruned.
+     */
+    final TipsViewModel tipsViewModel;
+
+    /**
+     * A map of {@link JobParser}s allowing us to determine how to parse the jobs from the garbage collector state file,
+     * based on their type.
+     */
+    private final HashMap<String, JobParser> jobParsers = new HashMap<>();
+
+    /**
+     * A map of {@link QueueProcessor}s allowing us to process queues based on the type of the job.
+     */
+    private final HashMap<Class<? extends GarbageCollectorJob>, QueueProcessor> queueProcessors = new HashMap<>();
+
+    /**
+     * A map of {@link QueueConsolidator}s allowing us to consolidate the queues to consume less "space".
+     */
+    private final HashMap<Class<? extends GarbageCollectorJob>, QueueConsolidator> queueConsolidators = new HashMap<>();
+
+    /**
+     * List of cleanup jobs that shall get processed by the garbage collector (grouped by their class).
+     */
+    private final HashMap<Class<? extends GarbageCollectorJob>, ArrayDeque<GarbageCollectorJob>> garbageCollectorJobs = new HashMap<>();
+
+    /**
+     * The constructor of this class stores the passed in parameters for future use and restores the previous state of
+     * the garbage collector if there is a valid one (to continue with cleaning up after IRI restarts).
+     *
+     * Before restoring the garbage collector state it registers the available job types, so we know how to parse and
+     * process the found jobs.
+     */
+    public GarbageCollector(Tangle tangle, SnapshotManager snapshotManager, TipsViewModel tipsViewModel) {
+        this.tangle = tangle;
+        this.snapshotManager = snapshotManager;
+        this.tipsViewModel = tipsViewModel;
+
+        MilestonePrunerJob.registerInGarbageCollector(this);
+        UnconfirmedSubtanglePrunerJob.registerInGarbageCollector(this);
+
+        restoreCleanupJobs();
+    }
+
+    /**
+     * This method adds a job to the GarbageCollector, that will get processed on its next run.
+     *
+     * It first registers the garbage collector in the job, so the job has access to all the properties that are
+     * relevant for its execution, and that do not get passed in via its constructor. Then it retrieves the queue based
+     * on its class and adds it there.
+     *
+     * After adding the job to its corresponding queue we try to consolidate the queue and persist the changes in the
+     * garbage collector state file.
+     *
+     * @param job the job that shall be executed
+     * @throws GarbageCollectorException if anything goes wrong while adding the job
+     */
+    public void addJob(GarbageCollectorJob job) throws GarbageCollectorException {
+        job.registerGarbageCollector(this);
+
+        ArrayDeque<GarbageCollectorJob> jobQueue = getJobQueue(job.getClass());
```

Comment:
In `SOLID` we have `D` = "depends upon abstractions not concretion".
It will be a good habit to write `Queue` instead of `ArrayDeque`. 

You use ArrayDeque in practice but you don't care about it, and it can be easily replaced.

Reply:
Since i need things like addLast and so on I kind of need to be specific here.